k8s/src/main/scala/io/buoyant/k8s/ServiceNamer.scala

```diff
@@ -18,16 +18,29 @@ import java.net.InetSocketAddress
  */
 class ServiceNamer(
   idPrefix: Path,
+  labelName: Option[String],
   mkApi: String => NsApi,
   backoff: Stream[Duration] = Backoff.exponentialJittered(10.milliseconds, 10.seconds)
 )(implicit timer: Timer = DefaultTimer.twitter) extends Namer {
 
   private[this] val PrefixLen = 3
+  private[this] val variablePrefixLength = if (labelName.isEmpty) PrefixLen else PrefixLen + 1
 
-  override def lookup(path: Path): Activity[NameTree[Name]] = path.take(PrefixLen) match {
-    case id@Path.Utf8(nsName, portName, serviceName) =>
+  def lookup(path: Path): Activity[NameTree[Name]] = (path.take(variablePrefixLength), labelName) match {
+    case (id@Path.Utf8(nsName, portName, serviceName), None) =>
+      val residual = path.drop(variablePrefixLength)
+      val nameTree = serviceNs.get(nsName, None).get(serviceName, portName).map {
+        case Some(address) =>
```

Comment:
since this is duplicated below you could probably factor it out into a `toNameTree` method